### DISCUSSION TITLE
Import and Render a GLTF model, and Iterate its Faces

### DISCUSSION THREAD

**Discussion by Bill-XU** [Open-Cascade-SAS/OCCT#95] [Q&A]
Hello,

I am using Python-OCC, which is a Python package for interfacing OCCT core functions, to deal with 3D models.
I use RWGltf and related packages to import GLTF models. The models can be correctly rendered in an OCCViewer.

But when I tried to select the model in OCCViewer, some error occurred.
And when I tried to iterate Face in the Shape, (I imported the model as a single shape), there is only one Face and without any Vertex attached on it.

Here is my code for import GLTF models:
```python
if extension in ('glb', 'gltf'):
        doc = TDocStd_Document("pythonocc-doc")
        gltf_reader = RWGltf_CafReader()
        gltf_reader.SetSystemCoordinateSystem(RWMesh_CoordinateSystem_Zup)
        gltf_reader.SetDocument(doc)
        status = gltf_reader.Perform(filename, Message_ProgressRange())
        assert status == IFSelect_RetDone
        shape = gltf_reader.SingleShape()
```

Here is how I iterate Face in a Shape:
```python
    explorer:TopExp_Explorer = TopExp_Explorer(shape, TopAbs_ShapeEnum.TopAbs_FACE)
    
    faces = []
    while explorer.More():
        faces.append(explorer.Value())
        explorer.Next()
```

Can anybody help explain the right way to deal with GLTF?
Or, how far does OCCT support GLTF?

Best regards,
Bill

**Comment by dpasukhi** [comment]
Hello.
Could you share the file?

OCCT reads the gltf with keeping the assembly and sub-shape structure. It means no combining into single mesh.
Natively OCCT do not have "SingleShape" method. In your case you put "SetDocument" that means you save all the shapes into "doc" (include assembly tree, names, colors, materials and other attributes).

So, if you will try that on OCCT. If file by default have multiple faces, you will receive them. I can't say nothing for the "SingleShape" method. IT is not a part of OCCT.

**Comment by Bill-XU** [comment]
Hello, @dpasukhi 

Here is the file. (I downloaded it for test purposes from an internet site)
[Duck.zip](https://github.com/user-attachments/files/17358042/Duck.zip)

I have checked source code behind SingleShape method.
It seems that SingleShape is a method of RWMesh_CafReader which is the parent class of RWGltf_CafReader.
(You can refer this: https://dev.opencascade.org/doc/refman/html/class_r_w_mesh___caf_reader.html)

According to this page: https://dev.opencascade.org/doc/overview/html/occt_user_guides__de_wrapper.html
I think the package RWGltf is for read/write GLTFs, but I can't find detailed description of how shapes and meshes are stored and how to iterate them.
Could you kindly give a hint where such information can be found?
Thank you!

Best regards.
Bill

**Reply by dpasukhi** [reply]
Yes, you are right, SingleShape is a part of RWMesh. I forgot about that because it is rudiment functionality (just to have symmetric with RWBRep).
As for a Gltf structure, OCCT have no CAD-depend structure. All DataExchange result saved as TopoDS_Shape or XCAF document.

As for a your file. You file indeed have only 1 face.  Please use a file that have a structure or at least well-know files.
_
parsing XCAF structure on python-occ must be simplified. Please check their implementation of graph iterator or walker or visitor or something developed on their side. I remember that they have.
The goal is - determinate type of label: assembly, simple shape, product, instance. And extract, collect attributes based on tree graph.

**Reply by Bill-XU** [reply]
@dpasukhi 
Thank you. I will check how XCAF structure is handled in python-occ.
